arch/x86/compare258_avx.c

```diff
@@ -0,0 +1,73 @@
+/* compare258_avx.c -- AVX2 version of compare258
+ * Copyright Mika T. Lindqvist  <postmaster@raasu.org>
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "../../zbuild.h"
+#include "../../zutil.h"
+
+#include "fallback_builtins.h"
+
+#if defined(X86_AVX2) && defined(HAVE_BUILTIN_CTZ)
+
+#include <immintrin.h>
+#ifdef _MSC_VER
+#  include <nmmintrin.h>
+#endif
+
+/* UNALIGNED_OK, AVX2 instrinic comparison */
+static inline int32_t compare256_unaligned_avx2_static(const unsigned char *src0, const unsigned char *src1) {
+    int32_t len = 0;
+    
+    do {
+        __m256i ymm_src0, ymm_src1, ymm_cmp;
+        ymm_src0 = _mm256_loadu_si256((__m256i*)src0);
+        ymm_src1 = _mm256_loadu_si256((__m256i*)src1);
+        ymm_cmp = _mm256_cmpeq_epi8(ymm_src0, ymm_src1); /* non-identical bytes = 00, identical bytes = FF */
+        int mask = _mm256_movemask_epi8(ymm_cmp); 
+        if ((unsigned int)mask != 0xFFFFFFFF) {
+            int match_byte = __builtin_ctz(~mask); /* Invert bits so identical = 0 */
+            return (int32_t)(len + match_byte);
+        }
+
+        src0 += 32, src1 += 32, len += 32;
+
+        ymm_src0 = _mm256_loadu_si256((__m256i*)src0);
+        ymm_src1 = _mm256_loadu_si256((__m256i*)src1);
+        ymm_cmp = _mm256_cmpeq_epi8(ymm_src0, ymm_src1);
+        mask = _mm256_movemask_epi8(ymm_cmp); 
+        if ((unsigned int)mask != 0xFFFFFFFF) {
+            int match_byte = __builtin_ctz(~mask);
+            return (int32_t)(len + match_byte);
+        }
+
+        src0 += 32, src1 += 32, len += 32;
+    } while (len < 256);
+
+    return len;
+}
+
+static inline int32_t compare258_unaligned_avx2_static(const unsigned char *src0, const unsigned char *src1) { 
+    if (*(uint16_t *)src0 != *(uint16_t *)src1)
+        return (*src0 == *src1);
+
+    src0 += 2, src1 += 2;
+    if (*src0 != *src1)
+        return 2;
+    if (src0[1] != src1[1])
+        return 3;
```

Comment:
8-bit compare is really slow, this should be changed to 16-bit compare like with first two bytes.

Reply:
Done.